pkl-cli/pkl-cli.gradle.kts

```diff
@@ -121,33 +130,95 @@ val testJavaExecutable by
+private fun setupJavaExecutableRun(
+  name: String,
+  args: Array<String>,
+  launcher: Provider<JavaLauncher>? = null,
+  configurator: Exec.() -> Unit = {},
+) =
+  tasks.register(name, Exec::class) {
     dependsOn(javaExecutable)
-    val outputFile =
-      layout.buildDirectory.file(
-        "testStartJavaExecutable"
-      ) // dummy output to satisfy up-to-date check
+    val outputFile = layout.buildDirectory.file(name) // dummy output to satisfy up-to-date check
     outputs.file(outputFile)
 
-    if (buildInfo.os.isWindows) {
-      executable = "java"
-      args("-jar", javaExecutable.get().outputs.files.singleFile.toString(), "--version")
-    } else {
-      executable = javaExecutable.get().outputs.files.singleFile.toString()
-      args("--version")
-    }
+    executable =
+      when (launcher) {
+        null -> "java"
+        else -> launcher.get().executablePath.asFile.absolutePath
+      }
+
+    args("-jar", javaExecutable.get().outputs.files.singleFile.toString(), *args)
 
     doFirst { outputFile.get().asFile.delete() }
 
     doLast { outputFile.get().asFile.writeText("OK") }
+
+    configurator()
+  }
+
+// 0.14 Java executable was broken because javaExecutable.jvmArgs wasn't commented out.
+// To catch this and similar problems, test that Java executable starts successfully.
+val testStartJavaExecutable by
+  setupJavaExecutableRun("testStartJavaExecutable", arrayOf("--version"))
+
+// Setup `testStartJavaExecutable` tasks for multi-JDK testing.
+val testStartJavaExecutableOnOtherJdks =
+  if (buildInfo.multiJdkTesting) {
+    buildInfo.jdkTestRange.map { jdkTarget ->
+      setupJavaExecutableRun(
+        "testStartJavaExecutableJdk${jdkTarget.asInt()}",
+        arrayOf("--version"),
+        serviceOf<JavaToolchainService>().launcherFor { languageVersion = jdkTarget },
+      )
+    }
+  } else {
+    emptyList()
   }
 
-tasks.check { dependsOn(testStartJavaExecutable) }
+val evalTestFlags = arrayOf("eval", "./.circleci/config.pkl")
+
+fun Exec.useRootDirAndSuppressOutput() {
+  workingDir = rootProject.layout.projectDirectory.asFile
+  standardOutput = ByteArrayOutputStream() // we only care that this exec doesn't fail
+}
+
+// 0.28 Preparing for JDK21 toolchains revealed that `testStartJavaExecutable` may pass, even though
+// the evaluator fails. To catch this, we need to test the evaluator. We render the CircleCI config
+// as a realistic test of the fat JAR.
+val testEvalJavaExecutable by
+  setupJavaExecutableRun("testEvalJavaExecutable", evalTestFlags) { useRootDirAndSuppressOutput() }
+
+// Run the same evaluator tests on all configured JDK test versions.
+val testEvalJavaExecutableOnOtherJdks =
+  buildInfo.jdkTestRange.map { jdkTarget ->
+    setupJavaExecutableRun(
+      "testEvalJavaExecutableJdk${jdkTarget.asInt()}",
+      evalTestFlags,
+      serviceOf<JavaToolchainService>().launcherFor { languageVersion = jdkTarget },
+    ) {
+      useRootDirAndSuppressOutput()
+    }
+  }
```

Comment:
These new tests effectively run:
```
jpkl eval ./.circleci/config.pkl > /dev/null
```
... to make sure that fat JARs run as expected end-to-end. This change is inspired by the `testStartJavaExecutableJdk17` passing even though GraalVM 17 ultimately refused to load the evaluator.

Same as all other multi-JDK tests, `jpkl eval ...` will be run on the toolchain JDK only unless multi-JDK testing is active.